src/twisted/trial/_asynctest.py

```diff
@@ -139,8 +112,39 @@ def onTimeout(d):
         else:
             d = defer.maybeDeferred(func)
 
-        call = self._twistedPrivateScheduler.callLater(timeout, onTimeout, d)
-        d.addBoth(lambda x: call.active() and call.cancel() or x)
+        if d.called or d.callbacks:
```

Comment:
No - this isn't how you use Deferred.

Reply:
That's why I marked the PR as draft :)

However, could we expose `Deferred.hasResult` or something like that which would give the caller information about whether there are any pending callbacks or not?

Reply:
Unfortunately, no.  It's fundamentally incompatible with the way Deferred works.  The way you get a result is you add a callback.  Any other approach risks receiving a different outcome than the APIs are supposed to produce.  In an application you might decide to accept this risk (perhaps because you know that in some case in your application it cannot result in a problem) but in trial where arbitrary application code will produce the Deferreds, it's not reasonable.

Reply:
It is absolutely a mistake that we exposed `.called` and `.callbacks` as public attributes in the first place, and something we should eventually correct.

Reply:
I fully agree that accessing `called` and `callbacks` is bad idea. The idea with `hasResult()` is to allow the caller to understand whether a adding an additional callback would result in it being called synchronously. User would still need to actually add a callback to fetch the result.

I expect that roughly half of the 30% performance improvement is due to having an equivalent of `hasResult()` and not doing additional unnecessary work.

Reply:
I've removed the change and the performance optimization has been reduced from 30% to 19% as tested by codspeed.